### PR TITLE
Added callbacks for DM, H, and S matrices export.

### PR DESCRIPTION
**Pull request by PavelStishenko**
Squashed commit of the following:

commit ddbd6d3e1065d8ca969d1791949b1c872cb21af2
Merge: b6c973df 407ad70e
Author: Pavel Stishenko <stishenkop@caridff.ac.uk>
Date:   Tue Apr 26 17:53:17 2022 +0100

    Merge remote-tracking branch 'upstream/main' into api-dm-2

    # Conflicts:
    #	external/s-dftd3/origin

commit b6c973dfba560303439bbedb803d34fc15080813
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Tue Mar 22 10:21:33 2022 +0000

    Minor cleanup

commit 878448f4e5ffde1445429a4bf8d69b4f2fe154af
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Sun Mar 20 21:41:08 2022 +0000

    Use straightforward association check

commit bda34f3a0c24958faead8a33b3ccc31ecf72680e
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Sun Mar 20 21:17:13 2022 +0000

    Fixed non-associated callbacks behaviour for GNU compiler

commit 62aaf2df79eafd5f8f39530e21cd0e872998c2ac
Merge: 36907aa8 199d8cf7
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Sun Mar 20 19:42:06 2022 +0000

    Merge remote-tracking branch 'upstream/main' into api-dm

    # Conflicts:
    #	src/dftbp/dftbplus/main.F90

commit 36907aa89d82c1d7f2d6af07d5e9237da662b393
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Sun Mar 20 19:27:36 2022 +0000

    Added Fortran layer for apicallback

commit 99c7cdabfef67451bdcc806696bad0006729177d
Merge: 65478059 618af2a7
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Tue Mar 15 10:45:19 2022 +0000

    Merge remote-tracking branch 'upstream/main' into api-dm

commit 65478059d04b6fb88d7da2525589e6f8e349f7fe
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Mon Feb 21 12:25:49 2022 +0000

    Fixed types in _autotest.tag

commit 12a31a6a873a4a256da8efb10212e1a36ee57556
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Sun Feb 20 22:11:40 2022 +0000

    added k-point and spin chanell indices to overlap and hamiltonian exporting callbacks

commit adbb2519b9f981259221632e38112c7c0f098114
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Thu Feb 17 14:43:23 2022 +0000

    Added comments to API-related code

commit 8b984cf47472631c43a876f8c9041aac186c588c
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Tue Feb 15 22:20:02 2022 +0000

    Extended test_mpiexpdmhs applicability

commit aaeebb298304ee6d83dfb25bcb13b12eabfac9b7
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Sun Feb 13 20:57:24 2022 +0000

    Added test_mpiexpdmhs

commit ecc8ef12d5185fe4748aaf4619a9b22517fbeab1
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Fri Feb 11 17:07:39 2022 +0000

    Added test for DM, H, S export with k-points

commit 575716db91a42b2c1c222eb187966653cb533b53
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Thu Feb 10 21:29:10 2022 +0000

    Revert "save test logs"

    This reverts commit 0768a1f13284c302e651972c8706abadcbc4175c.

commit f02db4982ef75d549ccb6d50a9d523f338adc8ca
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Thu Feb 10 21:28:30 2022 +0000

    Revert "Fixed build.yml"

    This reverts commit f54594779c7c92dac10c157d295c7f7a07ec7b88.

commit b1912c00fdecbe697ab8caa048c8bba0524c29ed
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Thu Feb 10 21:28:20 2022 +0000

    Revert "Upload logs in case of failed tests"

    This reverts commit 91e90098f828c7a4613f8ae8cf9e0b5c98197350.

commit c55dd38014cc302f1eb9303c643defae2ec9a54a
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Thu Feb 10 21:27:44 2022 +0000

    Fixed default initialization of TAPICallback members

commit 91e90098f828c7a4613f8ae8cf9e0b5c98197350
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Thu Feb 10 20:40:47 2022 +0000

    Upload logs in case of failed tests

commit f54594779c7c92dac10c157d295c7f7a07ec7b88
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Thu Feb 10 20:31:10 2022 +0000

    Fixed build.yml

commit 0768a1f13284c302e651972c8706abadcbc4175c
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Thu Feb 10 20:24:47 2022 +0000

    save test logs

commit 6f1157416aa88acfc6355605ba11155e43aaf231
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Thu Feb 10 19:38:24 2022 +0000

    Fixed kind of dftbp_is_hs_real resut

commit 7cbcffc80b5cc03d2377df4b9cead93418e0b06b
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Thu Feb 10 18:55:57 2022 +0000

    Fixed invokeS/H calls for complex case

commit 13a24ac22cedc07bd2f479eded866ef360ec3ef5
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Thu Feb 10 17:16:48 2022 +0000

    Code cleanup

commit cd4cdc85f427640e9d84d0f8c434ea72e0bd9990
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Thu Feb 10 16:48:02 2022 +0000

    Added support of builds without scalapack

commit 46634d81c563ff3d1fe8df1037a0c9d1f00821fb
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Thu Feb 10 16:47:38 2022 +0000

    code cleanup

commit db4a424ebe38b2a02c0a5d992ee9a62837a0bbe0
Merge: 014703d7 d6d4ea3d
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Thu Feb 10 16:46:57 2022 +0000

    Merge remote-tracking branch 'upstream/main' into api-dm

    # Conflicts:
    #	src/dftbp/api/mm/capi.F90
    #	src/dftbp/api/mm/mmapi.F90
    #	src/dftbp/dftbplus/mainapi.F90

commit 014703d7dde1208f0badc527ca28716c5d649283
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Wed Feb 9 23:11:26 2022 +0000

    Completed H & S export

commit 74b9737189f459d727b9cfe32c747e0fb4ce6d3c
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Thu Feb 3 23:28:22 2022 +0000

    added DM and S callbacks

commit 21c3cefc58f382a5db3ecc280cb04e394c86bd31
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Fri Jan 28 22:58:27 2022 +0000

    implemented dftbp_get_local_ks

commit 35947d5f00ba9920b5e65e217287b7499343d71a
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Mon Jan 24 22:46:43 2022 +0000

    Fixed setGeometry behaviour for calls without latVecs argument in PBC case

commit f11009b790afcfc78e6fd48f971203497d8e29a9
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Mon Jan 24 22:45:31 2022 +0000

    Implemented dftbp_get_nr_local_ks

commit 1291276b32da530d51ac29447906f38f825e5c9a
Author: Pavel Stishenko <stishenkop@cardiff.ac.uk>
Date:   Wed Jan 19 22:18:05 2022 +0000

    Added nSpin and nKpts functions to MM API

### CODE REVIEW

**Comment by PavelStishenko** [comment]
This pull requests replaces #956

**Comment by PavelStishenko** [comment]
@bhourahine I am sorry for making yet another PR, but I failed to rebase my branch so I had to create new branch and `git merge --squash` it with main. I hope it is OK. The commit message currently includes all comments of squashed commits, but it is still a single commit. If you think that old messages shouldn't be here, I will remove them.

**Comment by bhourahine** [comment]
Will put it through the buildbot and see what happens

File: src/dftbp/dftbplus/main.F90 [hunk 1]
```diff
@@ -2942,10 +2988,14 @@ subroutine buildAndDiagDenseCplxHam(env, denseDesc, ints, kPoint, neighbourList,
       else
         call unpackHS(HSqrCplx, ints%hamiltonian(:,iSpin), kPoint(:,iK), neighbourList%iNeighbour,&
             & nNeighbourSK, iCellVec, cellVec, denseDesc%iAtomStart, iSparseStart, img2CentCell)
-        call unpackHS(SSqrCplx, ints%overlap, kPoint(:,iK), neighbourList%iNeighbour, nNeighbourSK,&
-            & iCellVec, cellVec, denseDesc%iAtomStart, iSparseStart, img2CentCell)
+        call unpackHS(SSqrCplx, ints%overlap,              kPoint(:,iK), neighbourList%iNeighbour,&
```
Comment (bhourahine) [inline src/dftbp/dftbplus/main.F90:2991]:
```suggestion
        call unpackHS(SSqrCplx, ints%overlap, kPoint(:,iK), neighbourList%iNeighbour,&
```

**Review by bhourahine — COMMENTED** [review]
@PavelStishenko can you again squash and rebase? (there are conflicts on two files) 

I'm still getting regression fails on the buildbot (still looking at this).

@aradi c_ptrs now localised in API routines

**Comment by PavelStishenko** [comment]
> @PavelStishenko can you again squash and rebase? (there are conflicts on two files)
> 
I believe that I managed to do it.  Woohoo!

**Comment by bhourahine** [comment]
> I believe that I managed to do it. Woohoo!

Thanks. Not sure if we'll catch the 22.1 release, but aiming for 22.2.
There seems to be a problem with the tester being built with CC instead of MPICC, so missing various paths on some platforms (mpi.h not being found in particular).

**Comment by PavelStishenko** [comment]
> There seems to be a problem with the tester being built with CC instead of MPICC, so missing various paths on some platforms (mpi.h not being found in particular).

I configured MPI build this way:
`LDFLAGS="-Wl,--copy-dt-needed-entries" FC=mpifort CC=mpicc FFLAGS="-w" cmake -DLAPACK_LIBRARY=openblas -DSCALAPACK_LIBRARY=scalapack-openmpi -DCMAKE_INSTALL_PREFIX=${HOME}/opt/dftbp-mpi -B _build-mpi -C mpi.config.cmake .`
Probably I didn't found that bug because of this part: ` CC=mpicc`

File: test/src/dftbp/api/mm/testers/CMakeLists.txt [hunk 2]
```diff
@@ -14,6 +14,16 @@ if(WITH_C_EXECUTABLES)
   add_executable(test_fileinitc test_fileinitc.c $<TARGET_OBJECTS:test_api_mm_utils>)
   target_link_libraries(test_fileinitc dftbplus)
   set_target_properties(test_fileinitc PROPERTIES LINKER_LANGUAGE C)
+
+  add_executable(test_expdmhs test_expdmhs.c $<TARGET_OBJECTS:test_api_mm_utils>)
+  target_link_libraries(test_expdmhs dftbplus)
+  set_target_properties(test_expdmhs PROPERTIES LINKER_LANGUAGE C)
+  
+  if(WITH_MPI)
```
[the remark this answers is not in the document]
Comment (bhourahine) [reply test/src/dftbp/api/mm/testers/CMakeLists.txt:22]: Unfortunately still not resolving the MPICC behaviour.

File: test/src/dftbp/api/mm/testers/CMakeLists.txt [hunk 3]
```diff
@@ -14,6 +14,17 @@ if(WITH_C_EXECUTABLES)
...
+    find_package(MPI REQUIRED)
```
Comment (bhourahine) [inline test/src/dftbp/api/mm/testers/CMakeLists.txt:23]: `
  include_directories(SYSTEM ${MPI_INCLUDE_PATH})
`

File: test/src/dftbp/api/mm/testers/CMakeLists.txt [hunk 4]
```diff
@@ -14,6 +14,17 @@ if(WITH_C_EXECUTABLES)
...
+    add_executable(test_mpiexpdmhs test_mpiexpdmhs.c $<TARGET_OBJECTS:test_api_mm_utils>)
+    target_link_libraries(test_mpiexpdmhs dftbplus)
```
Comment (bhourahine) [inline test/src/dftbp/api/mm/testers/CMakeLists.txt:25]: `
target_link_libraries(test_mpiexpdmhs dftbplus ${MPI_C_LIBRARIES})
`
Possibly

Comment (aradi) [reply test/src/dftbp/api/mm/testers/CMakeLists.txt:25]: I think, it should be rather
```
target_link_libraries(... MPI::MPI_C)
```
As the target `dftbplus` only depends on `MPI::MPI_Fortran`, the C-libraries are probably indeed missing here.

@PavelStishenko Yes, if you set `CC=mpicc`, then it works out of the box. However, it should also work with `CC=gcc` and then CMake (hopefully) finding the right MPI-dependencies.

Comment (PavelStishenko) [reply test/src/dftbp/api/mm/testers/CMakeLists.txt]: `tests_mpiexpdmhs.c` pass on my machine. 
I use this command for cmake config:
`LDFLAGS="-Wl,--copy-dt-needed-entries" FC=gfortran-8 CC=gcc-8 FFLAGS="-w" cmake -DLAPACK_LIBRARY=openblas -DSCALAPACK_LIBRARY=scalapack-openmpi -DCMAKE_INSTALL_PREFIX=${HOME}/opt/dftbp-mpi2 -B _build-mpi2 -C mpi.config.cmake .`

And this config `mpi.config.cmake`: https://pastebin.com/9rELuLU5

**Comment by PavelStishenko** [comment]
> OK, it seems we neglected this PR for a very long time (sorry!) @PavelStishenko is that still the PR we should look at?

Yes, it is neglected. Current version is in PR #1217 . I am closing this one.

**Pull request closed without merging**